### PR TITLE
Fixes #14924: Cleanup unreferenced software 

### PR DESCRIPTION
**Pull request by ncharles**
https://issues.rudder.io/issues/14924

### CODE REVIEW

File: webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/ReadOnlySoftwareInventoryDAOImpl.scala [hunk 1]
```diff
@@ -90,4 +96,95 @@ class ReadOnlySoftwareDAOImpl(
...
+                             val ids: Seq[String] = softwareEntry.flatMap(entry => entry.valuesFor(A_SOFTWARE_DN).toSet )
+                             val results = sequence(ids) { id => acceptedDit.SOFTWARE.SOFT.idFromDN(new DN(id)) }
+
+                             results match {
+                               case Full(softIds) =>
+                                 mutSetSoftwares = mutSetSoftwares.map(t => t ++ softIds)
+                                 Full(Unit)
+                               case Failure(msg, exception, chain) => mutSetSoftwares = Failure(msg, exception, chain) // otherwise the time is wrong
+                             }
+                           }
```
Comment (ncharles) [inline webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/ReadOnlySoftwareInventoryDAOImpl.scala]: this ain't pretty, yet its more efficient
[2019-05-23 13:00:06] DEBUG com.normation.inventory.ldap.core.SoftwareServiceImpl - All softwares id in nodes fetched: 37953 softwares id in 10029ms, compared to what is in the TODO below
[2019-05-23 12:55:17] DEBUG com.normation.inventory.ldap.core.SoftwareServiceImpl - All softwares id in nodes fetched: 37953 softwares id in 15197ms

Comment (ncharles) [reply webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/ReadOnlySoftwareInventoryDAOImpl.scala]: this code is surely non idiomatic, and i fail to improve it

**Comment by ncharles** [comment]
Commit modified

File: webapp/sources/rudder/rudder-web/src/main/resources/logback.xml [hunk 2]
```diff
@@ -146,7 +146,7 @@ along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
     are doing.
   -->
 
-  <root level="info">
+  <root level="debug">
```
Comment (fanf) [inline webapp/sources/rudder/rudder-web/src/main/resources/logback.xml]: That should not be modified

File: webapp/sources/rudder/rudder-web/src/main/resources/logback.xml [hunk 3]
```diff
@@ -360,7 +360,7 @@ along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
     ================
     This logger is in charge of all scheduled jobs and batches
   -->
-  <logger name="scheduledJob" level="info" additivity="false">
+  <logger name="scheduledJob" level="debug" additivity="false">
```
Comment (fanf) [inline webapp/sources/rudder/rudder-web/src/main/resources/logback.xml]: neither that

**Comment by ncharles** [comment]
Commit modified

File: webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/ReadOnlySoftwareInventoryDAOImpl.scala [hunk 4]
```diff
@@ -90,4 +96,98 @@ class ReadOnlySoftwareDAOImpl(
...
+    val t1 = System.currentTimeMillis
+    for {
+      con           <- ldap
+
+      // fetch all nodes
+      nodes         = con.searchSub(acceptedDit.NODES.dn.getParent, IS(OC_NODE), A_NODE_UUID)
+
+      batchedNodes = nodes.grouped(50)
+
+      _ = batchedNodes.foreach { nodeEntries: Seq[LDAPEntry] =>
```
Comment (fanf) [inline webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/ReadOnlySoftwareInventoryDAOImpl.scala]: it should not be a foreach, else we loose the maybe `Failure`:  ```res <- sequence(batchedNodes) { .... }```

Comment (ncharles) [reply webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/ReadOnlySoftwareInventoryDAOImpl.scala]: i tried using a sequence on batchedNodes, without success (it's an iterator, I cannot sequence it)

Comment (fanf) [reply webapp/sources/ldap-inventory/inventory-repository/src/main/scala/com/normation/inventory/ldap/core/ReadOnlySoftwareInventoryDAOImpl.scala]: You can do:
``
      batchedNodes         = con.searchSub(acceptedDit.NODES.dn.getParent, IS(OC_NODE), A_NODE_UUID).grouped(50).toSeq

      res <- sequence(batchedNodes) { nodeEntries: Seq[LDAPEntry] =>
```
You still process nodes 50 by 50, and don't hold more node ids than previously

**Comment by ncharles** [comment]
Commit modified

**Comment by fanf** [comment]
OK, merging this PR

**Pull request merged**